program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,371 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    ///initializes a zero initialized tracker
+    fn initialize(&mut self) -> Result<(), OracleError>;
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError>;
+    /// returns the minimum number of entries that needs to be added
+    /// to the current entry of the time_machine so that the following two
+    /// conditions are satisfied
+    /// 1. after adding the output, the new current entry has the same offset as if
+    ///     the price was constantly updating
+    /// 2. any entry that would have been passed through during this time interval, would be
+    ///     passed through if we skip one entry at a time the number that this outputs
+    fn get_num_skipped_entries_capped(
+        prev_time: i64,
+        current_time: i64,
+    ) -> Result<usize, OracleError> {
+        let time_since_begining_of_current_entry =
+            (prev_time % GRANUALITY) + current_time - prev_time;
+        let mut num_skiped_entries: usize =
+            try_convert(time_since_begining_of_current_entry / GRANUALITY)?;
+        //if num_skipped is greater than NUM_ENTRIES, cap it at the smallest
+        //number greater than NUM_ENTRIES that has the same residue class
+        if num_skiped_entries >= NUM_ENTRIES {
```

Comment:
If `num_skiped_entries >= NUM_ENTRIES` you're just invalidating everything so I'd be OK to just set num_skiped_entries = NUM_ENTRIES

Reply:
That would break the wormhole thing, and the predicatiblity of `current_entry`. Though it would work with the next comment of ignoring `current_entry`